phabricatoremails/mail.py

```diff
@@ -0,0 +1,153 @@
+    def __init__(self, from_address: str, logger: Logger, output_path: pathlib.Path):
+        self._logger = logger
+        self._from_address = from_address
+        self._index = 0
+
+        self._output_path = output_path
+        self._eml_path = output_path / "eml"
+        self._html_path = output_path / "html"
+        self._text_path = output_path / "text"
+        self._eml_path.mkdir(parents=True, exist_ok=True)
+        self._html_path.mkdir(parents=True, exist_ok=True)
+        self._text_path.mkdir(parents=True, exist_ok=True)
+
+    def send(self, emails: List[OutgoingEmail]):
+        for email in emails:
+            basefilename = f"{self._index}-to-{email.to}"
+            with (self._eml_path / (basefilename + ".eml")).open("w") as file:
+                file.write(email.to_mime_message(self._from_address).as_string())
+            with (self._html_path / (basefilename + ".html")).open("w") as file:
+                file.write(email.html_contents)
+            with (self._text_path / (basefilename + ".text")).open("w") as file:
+                file.write(email.text_contents)
+
+            self._index += 1
+        self._logger.debug(
+            f'Recorded {len(emails)} emails to the "{self._output_path}" directory'
+        )
+
+
+@dataclass
+class SmtpMail:
+    """Sends emails via SMTP
+
+    This is useful for end-to-end testing emails in real email clients when an
+    SMTP server is available for use, but an Amazon SES account is not.
+    """
+
+    _server: smtplib.SMTP
+    _from_address: str
+    _logger: Logger
+    _send_to: Optional[str]
+
+    def send(self, emails: List[OutgoingEmail]):
+        for email in emails:
+            self._logger.debug(
+                f'[{email.to}] Sending "{email.template_name}" for "{email.subject}"'
+            )
+
+            mime_message = email.to_mime_message(
+                self._from_address, include_target_in_subject=self._send_to is not None
+            )
+            self._server.sendmail(
+                self._from_address,
+                self._send_to if self._send_to else email.to,
+                mime_message.as_string(),
+            )
+
+
+@dataclass
+class SesMail:
+    """Sends emails via Amazon SES"""
+
+    _client: Any
+    _from_address: str
+    _logger: Logger
+    _send_to: Optional[str]
+
+    @staticmethod
+    def from_aws_credentials(
+        from_address: str,
+        logger: Logger,
+        send_to: Optional[str],
+        aws_access_key_id: Optional[str],
+        aws_secret_access_key: Optional[str],
+    ):
+        """Automatically creates an SES client
+
+        Uses aws credentials either from parameters, or from the environment.
+        """
+        client = boto3.client(
+            "ses",
+            aws_access_key_id=aws_access_key_id,
+            aws_secret_access_key=aws_secret_access_key,
+        )
+        return SesMail(client, from_address, logger, send_to)
+
+    def send(self, emails: List[OutgoingEmail]):
```

Comment:
Hmm, I'm not sure that a docstring here would be valuable.
We're within `SesMail`, and we already know that the class's purpose is to "Send emails via Amazon SES" (according to the class-level docstring).

This is a `send()` method that takes a list of "outgoing emails".
Adding an explicit docstring for this method that it will "Send a raw email via the Amazon SES API." seems redundant.

Reply:
> how it's impacted by `send_to`

This part is documented where the class is created and `send_to` is set.

> the fact that it is using a particular type of SES API (i.e. `send_raw_email` and not `send_email`)

Good point about this, I'll add an inline comment about this, though it doesn't belong as a method docstring.

Reply:
Inline comments help the developer when the developer is looking through the code and trying to understand the code flow, assumptions made, etc. Docstrings should help explain what a method does and how it does it on a high level, and they're a useful tool when debugging. One would be able to invoke the built-in help system to try and understand how this method works where it is being used (e.g. if someone is in a breakpoint in`Pipeline.run` trying to debug something.) Given that these methods do different things in `SesMail`, `SmtpMail`, etc., they should have docstrings.

Reply:
> Given that these methods do different things in SesMail, SmtpMail, etc., they should have docstrings.

The different things that these methods do is directly tied to the class name: `SesMail.send()` calls SES, `SMTPMail.send()` uses SMTP, `FSMail` dumps to the file system. This is documented for each class.

The specifics of how this works (e.g.: `SESMail` calling `send_raw_email` instead of `send_email`) is an _implementation detail_. Callers of `<impl Mail>.send()` don't need to worry about it.

> Docstrings should help explain what a method does and how it does it on a high level

Docstrings should explain what a method does and perhaps how parameters affects its behaviour, sure. They shouldn't explain explain implementation details.

> One would be able to invoke the built-in help system to try and understand how this method works where it is being used (e.g. if someone is in a breakpoint in `Pipeline.run` trying to debug something.)

If someone is debugging and they want to see how a class functions, they can step into methods if they need to see implementation details.